File: ModelFitting/hmm_viterbi.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
From the wikipedia page with slight modification
https://en.wikipedia.org/wiki/Viterbi_algorithm#Example
"""

def viterbi(obs, states, start_p, trans_p, emit_p):
    V = [{}]

    for st in states:
        V[0][st] = {"prob": start_p[st] * emit_p[st][obs[0]], "prev": None}

    # Run Viterbi when t > 0

    for t in range(1, len(obs)):
        V.append({})

        for st in states:
            max_tr_prob = max(V[t-1][prev_st]["prob"]*trans_p[prev_st][st] for prev_st in states)

            for prev_st in states:
                if V[t-1][prev_st]["prob"] * trans_p[prev_st][st] == max_tr_prob:
                    max_prob = max_tr_prob * emit_p[st][obs[t]]
                    V[t][st] = {"prob": max_prob, "prev": prev_st}
                    break

    for line in dptable(V):
        print(line)

    opt = []

    # The highest probability
    max_prob = max(value["prob"] for value in V[-1].values())
    previous = None

    # Get most probable state and its backtrack
    for st, data in V[-1].items():
        if data["prob"] == max_prob:
            opt.append(st)
            previous = st
            break

    # Follow the backtrack till the first observation
    for t in range(len(V) - 2, -1, -1):
        opt.insert(0, V[t + 1][previous]["prev"])
        previous = V[t + 1][previous]["prev"]

    print('The steps of states are ' + ' '.join(opt) + ' with highest probability of %s' % max_prob)


def dptable(V):
    # Print a table of steps from dictionary
    yield " ".join(("%12d" % i) for i in range(len(V)))
    
    for state in V[0]:
        yield "%.7s: " % state + " ".join("%.7s" % ("%f" % v[state]["prob"]) for v in V)

#The function viterbi takes the following arguments: obs is the sequence of observations, e.g. ['normal', 'cold', 'dizzy']; states is the set of hidden states; start_p is the start probability; trans_p are the transition probabilities; and emit_p are the emission probabilities. For simplicity of code, we assume that the observation sequence obs is non-empty and that trans_p[i][j] and emit_p[i][j] is defined for all states i,j.

#In the running example, the forward/Viterbi algorithm is used as follows:


obs = ('normal', 'cold', 'dizzy')
states = ('Healthy', 'Fever')
start_p = {'Healthy': 0.6, 'Fever': 0.4}
trans_p = {
   'Healthy' : {'Healthy': 0.7, 'Fever': 0.3},
   'Fever' : {'Healthy': 0.4, 'Fever': 0.6}
   }
emit_p = {
   'Healthy' : {'normal': 0.5, 'cold': 0.4, 'dizzy': 0.1},
   'Fever' : {'normal': 0.1, 'cold': 0.3, 'dizzy': 0.6}
   }
viterbi(obs,states,start_p,trans_p,emit_p)

File: ModelFitting/hmm_viterbi.R

# Description -------------------------------------------------------------

# This function duplicates hmm_viterbi.py, which comes from the Viterbi
# algorithm wikipedia page. This first function is just to provide R code that
# is similar in case anyone is interested in comparison, but the original used
# lists of tuples and thus was very inefficient and provided output that wasn't
# succinct.  The second function takes a vectorized approach and returns a
# matrix in a much more straightforward fashion.  Both will provide the same
# result as the Python code.



# Original in R -----------------------------------------------------------

viterbi <- function(obs, states, start_p, trans_p, emit_p) {
  V = vector('list', length(obs))
  
  for (st in seq_along(states)) {
    V[[1]][[states[st]]] = list("prob"= start_p[st] * emit_p[[st]][obs[1]], "prev"= NULL)
  }
  
  for (t in 2:length(obs)) {
    
    for (st in seq_along(states)) {
      max_tr_prob = numeric()
      
      for (prev_st in states) {
        max_tr_prob[prev_st] = V[[t-1]][[prev_st]][["prob"]] * trans_p[[prev_st]][[st]]
      }
      max_tr_prob = max(max_tr_prob)
      
      for (prev_st in states) {
        flag =  V[[t-1]][[prev_st]][["prob"]] * trans_p[[prev_st]][[st]] == max_tr_prob
        if (flag) {
          max_prob = max_tr_prob * emit_p[[st]][obs[t]]
          V[[t]][[states[st]]] = list('prob' = max_prob, 'prev' = prev_st)
        }
      }
    }
  }
  
  # I don't bother duplicating the text output code
  df_out = rbind(Healthy = sapply(V, function(x) x$Healthy$prob),
                 Fever = sapply(V, function(x) x$Fever$prob))
  colnames(df_out) = obs
  print(df_out)
  m = paste0('The steps of states are: ', 
             paste(rownames(df_out)[apply(df_out, 2, which.max)], collapse = ' '), 
             paste('\nHighest probability: ', max(df_out[,ncol(df_out)])))
  message(m)
  V
}


#: Data Setup --------------------------------------------------------------

obs = c('normal', 'cold', 'dizzy')
states = c('Healthy', 'Fever')
start_p = c('Healthy'= 0.6, 'Fever'= 0.4)
trans_p = list(
  'Healthy' = c('Healthy'= 0.7, 'Fever'= 0.3),
  'Fever' = c('Healthy'= 0.4, 'Fever'= 0.6)
)
emit_p = list(
  'Healthy' = c('normal'= 0.5, 'cold'= 0.4, 'dizzy'= 0.1),
  'Fever' = c('normal'= 0.1, 'cold'= 0.3, 'dizzy'= 0.6)
)


#: Demo --------------------------------------------------------------------

test = viterbi(obs, states, start_p, trans_p, emit_p)
# test

set.seed(123)
obs = sample(obs, 6, replace = T)
test = viterbi(obs, states, start_p, trans_p, emit_p)
# test


# Vectorized --------------------------------------------------------------

viterbi_sane <- function(obs, states, start_p, trans_mat, emit_mat) {
  prob_mat = matrix(NA, nrow = length(states), ncol = length(obs))
  colnames(prob_mat) = obs
  rownames(prob_mat) = states

  prob_mat[,1] = start_p * emit_mat[,1]

  for (t in 2:length(obs)) {
    prob_tran = prob_mat[,t-1] * trans_mat
    max_tr_prob =  apply(prob_tran, 2, max)
    prob_mat[,t] =  max_tr_prob * emit_mat[, obs[t]]
  }
  
  print(prob_mat)
  m = paste0('The steps of states are: ', 
             paste(states[apply(prob_mat, 2, which.max)], collapse = ' '), 
             paste('\nHighest probability: ', max(prob_mat[,ncol(prob_mat)])))
  message(m)
}


#: Data Setup --------------------------------------------------------------

# pick data
obs = c('normal', 'cold', 'dizzy')
set.seed(123)
obs = sample(obs, 6, replace = T)

# need matrices now
emit_mat = do.call(rbind, emit_p)
trans_mat = do.call(rbind, trans_p)


#: Demo --------------------------------------------------------------------

viterbi_sane(obs, states, start_p, trans_mat, emit_mat)



# A final demo ------------------------------------------------------------

# from the hidden markov model wikipedia page

states = c('Rainy', 'Sunny')

observations = c('walk', 'shop', 'clean')

start_probability = c('Rainy'= 0.6, 'Sunny'= 0.4)

transition_probability = rbind(
  'Rainy' = c('Rainy'= 0.7, 'Sunny'= 0.3),
  'Sunny' = c('Rainy'= 0.4, 'Sunny'= 0.6)
)

emission_probability = rbind(
  'Rainy' = c('walk'= 0.1, 'shop'= 0.4, 'clean'= 0.5),
  'Sunny' = c('walk'= 0.6, 'shop'= 0.3, 'clean'= 0.1)
)

viterbi_sane(observations, states, start_probability, transition_probability, emission_probability)
